src/metabase/models/user.clj

```diff
@@ -82,10 +82,18 @@
 (defn- pre-update
   [{reset-token :reset_token, superuser? :is_superuser, active? :is_active, :keys [email id locale], :as user}]
   ;; when `:is_superuser` is toggled add or remove the user from the 'Admin' group as appropriate
-  (when (some? superuser?)
-    (let [membership-exists? (db/exists? PermissionsGroupMembership
-                               :group_id (:id (perms-group/admin))
-                               :user_id  id)]
+  (let [membership-exists? (db/exists? PermissionsGroupMembership
+                             :group_id (:id (perms-group/admin))
+                             :user_id  id)
+        still-superuser?   (or superuser?
+                               (and (nil? superuser?) membership-exists?))]
+    ;; Do not let the last admin archive themselves
+    (when (and still-superuser?
+               (false? active?)
+               (= 1 (perms-group-membership/admin-count))) ;; The 1 is this very user
```

Comment:
Why isn't this sufficient? I'm having trouble imagining why `superuser?`'s value has any relevance here.
```suggestion
    (when (and membership-exists?
               (false? active?)
               (= 1 (perms-group-membership/admin-count))) ;; The 1 is this very user
```

Reply:
There's a scenario (probably unreachable from the current API, but I want to be future-proof since this is model-level code) where a user is not an admin, then in one update they become an admin and also get archived. In that scenario, we'll have:

| var | value
| -- | -- |
| `membership-exists?` | `false` |
| `active?` | `false` |
| `superuser?` | `true` |
| `still-superuser?` | `true` |

Reply:
In that scenario, since the user is not an admin, there must be at least one other admin. So setting this user to be admin and also archiving them should be a valid operation, albeit senseless.